File: HackerRank/Contests/Math-Ad-Infinitum-13/multiply_moderate.hs
-- -------------------------------------------------------------------------------------
--         Author: Sourabh S Joshi (cbrghostrider); Copyright - All rights reserved.
--                       For email, run on linux (perl v5.8.5):
--   perl -e 'print pack "H*","736f75726162682e732e6a6f73686940676d61696c2e636f6d0a"'
-- -------------------------------------------------------------------------------------

-- The Haskell version is slower, but still correct
-- Including only for completeness

import Data.List

-------------------------------------------------------------------------------
-- Credit : The following (efficient) sieve is not my code
primels :: [Integer]
primels = 2 : filter (isPrime primels) [3,5..]
  where isPrime (p:ps) n = p*p > n || n `rem` p /= 0 && isPrime ps n
-- End credit.        
-------------------------------------------------------------------------------

type PFactor = (Integer, Integer)

c_MODNUM :: Integer
c_MODNUM = 1000000007

getContribution :: PFactor -> Integer
getContribution (pr, ex) = (pr ^ (ex + 1) - 1) `div` (pr - 1)

sumProperDivisors :: [PFactor] -> Integer
sumProperDivisors = foldl' (\ acc elem -> acc * getContribution elem) 1

combineFacts :: [PFactor] -> [PFactor] -> [PFactor]
combineFacts [] ys = ys
combineFacts xs [] = xs
combineFacts ((x@(px, ex)):xs) ((y@(py, ey)):ys) 
  | px == py  = (px, ex + ey) : combineFacts xs ys
  | px < py   = x : combineFacts xs (y:ys)
  | otherwise = y : combineFacts (x:xs) ys

countTimes :: Integer -> Integer -> Integer
countTimes n p = if (n `mod` p == 0) then 1 + countTimes (n `div` p) p else 0

countPrimeTimes :: Integer -> Integer -> (Integer, PFactor)
countPrimeTimes n p = (newn, factor) 
  where factor = (p, countTimes n p)
        newn   = n `div` (p ^ (snd factor))

extractPrimeFact :: [Integer] -> Integer -> [PFactor]
extractPrimeFact [] _ = []
extractPrimeFact (p:ps) num 
  | num <= 1       = []
  | p > num        = []
  | (num `mod` p) == 0 = let (nn, pf) = countPrimeTimes num p in (pf : extractPrimeFact ps nn)
  | otherwise      = extractPrimeFact ps num

runQueries :: Int -> [Integer] -> [PFactor] -> IO ()
runQueries 0 _ _ = return ()
runQueries q primes oldpfs = do
  xstr <- getLine
  let x = read xstr
  let xpfs = extractPrimeFact primes x
  let newpfs = combineFacts oldpfs xpfs
  let ans = (sumProperDivisors newpfs) `mod` c_MODNUM
  putStrLn $ show ans
  runQueries (q-1) primes newpfs 


main :: IO ()
main = do 
  nqstr <- getLine 
  let allPrimes = take 10000 primels
  let [nstr, qstr] = words nqstr
      n = read nstr
      q = read qstr
  let oldpfs = extractPrimeFact allPrimes n
  runQueries q allPrimes oldpfs
  

File: HackerRank/Contests/Math-Ad-Infinitum-13/multiply_moderate.cpp
// -------------------------------------------------------------------------------------
//         Author: Sourabh S Joshi (cbrghostrider); Copyright - All rights reserved.
//                       For email, run on linux (perl v5.8.5):
//   perl -e 'print pack "H*","736f75726162682e732e6a6f73686940676d61696c2e636f6d0a"'
// -------------------------------------------------------------------------------------
#include <iostream>
#include <cstdio>
#include <vector>
#include <math.h>
#include <algorithm>
#include <unordered_map>
#include <utility>

using std::unordered_map;
using std::pair;
using std::make_pair;
using std::vector;

typedef unsigned long long Val;

#define NUMLIMIT (1000050)
#define MODNUM (1000000007)

unordered_map<Val, Val> primePowCache;

std::vector<Val> primes ;

inline Val makeKey(Val prime, Val exp) {
    return ((prime << 32) | (exp));
}

Val myPow(Val x, Val n) {
    if (n == 0) return 1;
    if (n == 1) return x;
    Val subPow = myPow(x, n/2); //divide and conquer
    Val retval = subPow * subPow;
    retval %= MODNUM;
    retval *= (n % 2 == 1 ? x : 1);
    retval %= MODNUM;
    return retval;
}

void generatePrimes(int n){
    std::vector<bool>sieve(n, true) ;
    for( std::size_t i = 2 ; i < sieve.size() ; ++i ) 
        if( sieve[i] ) {
            primes.push_back(i);
            
            //std::cout << "Prime: " << i << std::endl;
            for( std::size_t j = i*2 ; j < sieve.size() ; j += i ) {
                sieve[j] = false ;
            }
        }
}

struct PFactor {
    int prime; 
    int exp;
    PFactor(int p, int e) : prime(p), exp(e) {}

    //sum e = 0 to exp of (prime ^e)
    Val getContribution() const {
        Val ppc;
        auto it = primePowCache.find(prime);
        if (it == primePowCache.end()) {
            ppc = myPow(prime-1, MODNUM-2);
            primePowCache.insert(make_pair(prime, ppc));
        } else {
            ppc = it->second;
        }
        Val ans = (myPow(prime, exp+1) + MODNUM - 1) * (ppc);
        ans %= MODNUM;

        return ans;
    }
};

PFactor extractPrime(Val& num, Val prime) {
    PFactor ret(prime, 0);
    while ((num % prime) == 0) {
        num /= prime;
        ret.exp++;
    }
    return ret;
}

vector<PFactor> extractPrimeFact(Val num) {
    vector<PFactor> retval;
    int pind = 0;
    while (primes[pind] <= num) {
        //std::cout << "Checking " << primes[pind] << std::endl;
        if (num % primes[pind]) {
            pind++;
            continue;
        }

        PFactor pf = extractPrime(num, primes[pind++]);
        //std::cout << "Found pf: (" << pf.prime << ", " << pf.exp << ")" << std::endl;
        retval.push_back(pf);
    }

    return retval;
}

Val sumProperDivisors(const vector<PFactor>& pfs) {
    Val ret = 1;
    for (const PFactor& pf : pfs) {
        ret *= pf.getContribution();
        ret %= MODNUM;
    }
    return ret;
}

vector<PFactor> combineFacts(vector<PFactor>& oldpfs, vector<PFactor>& newpfs) {
    vector<PFactor> ret;
    int oi=0, ni=0;
    while (oi < oldpfs.size() && ni < newpfs.size()) {
        if (oldpfs[oi].prime == newpfs[ni].prime) {
            PFactor ppf(oldpfs[oi].prime, oldpfs[oi].exp + newpfs[ni].exp);
            ret.push_back(ppf);
            oi++; ni++;
        } else if (oldpfs[oi].prime < newpfs[ni].prime) {
            ret.push_back(oldpfs[oi++]);
        } else {
            ret.push_back(newpfs[ni++]);
        }
    }
    if (oi < oldpfs.size()) {
        ret.insert(ret.end(), oldpfs.begin() + oi, oldpfs.end());
    }
    if (ni < newpfs.size()) {
        ret.insert(ret.end(), newpfs.begin() + ni, newpfs.end());
    }
    return ret;
}

int main() {
    generatePrimes(NUMLIMIT);

    Val n; std::cin >> n;
    int q; std::cin >> q;

    vector<PFactor> oldpfs = extractPrimeFact(n);

    for (int qi=0; qi<q; qi++) {
        Val x; std::cin >> x;

        vector<PFactor> newpfs = extractPrimeFact(x);
        oldpfs = combineFacts(oldpfs, newpfs);
        Val ans = sumProperDivisors(oldpfs);

        std::cout << ans << std::endl;
    }
}
